### ISSUE TITLE
Allow spaces in type annotations for vertical alignment since () already serve as delimiters

### ISSUE THREAD

**Issue by eugenesvk** [kdl-org/kdl#367]
Example: let's say you have a keybinding config for a modal editor where you want to pass a mode for a keybind in types. The best way to allow easy visual differentiation of what mode a keybind belongs to would be to have the same mode always positioned in the same column like so:
```
(niv)Escape "normal_mode" // works in Normal/Insert/Visual modes
( i )Ctrl+I "normal_mode" // obvious that only one mode matches
// and that mode is Insert since 'i' is in the same spot
// and eyeballing for a diff is easier
```

But currently spaces are not allowed in types even though they aren't needed: you already have `()` to signal limits to the type (banning spaces in regular bare identifiers makes sense since spaces are used to separate different values)

Currently there is a workaround: use quotes, but for short types that would be using more symbols for delimitation than for content: `("")`
```
("niv")Escape "normal_mode"
(" i ")Ctrl+I "normal_mode"
```

**Comment by IceDragon200** [comment]
Spaces in type annotations [are allowed in kdl v2](https://github.com/kdl-org/kdl/blob/kdl-v2/SPEC.md#type-annotation).

So keep an eye on the PR #286

**Comment by eugenesvk** [comment]
That's better, though as far as I understood it would not allow whitespace **within** type values (since unquoted strings do not allow whitespace), so this would still be illegal
```
(niv)Escape "normal_mode"
(n v)Ctrl+I "normal_mode"
```

**Comment by IceDragon200** [comment]
Type annotations are not meant to be used in that manner, it denotes one value which "suggests" what the type should be.

Rather I believe if you flip the document around you achieve something slightly more tolerable
```kdl
(normal_mode)Escape niv
(normal_mode)Ctrl+I n v
```

Where the type annotation now denotes the mode and the arguments denote the key bindings
This has the benefit of allowing you to annotate the keys themselves now:
```kdl
(normal_mode)Escape (upper-only)niv
(normal_mode)Ctrl+I n (icase)v
```

You could also use sub nodes instead to denote the modes to remove the visual clutter of the annotations

```kdl
normal_mode {
  Escape niv
  Ctrl+I n v
}
```

**Comment by eugenesvk** [comment]
> Type annotations are not meant to be used in that manner, it denotes one value which "suggests" what the type should be.

A combination of modes is a type for a keybind! It's just that given the already existing delimiters `()` I thought we could get more flexibility by basically allowing (almost all of ) quoted strings without quotes (excluding `()`)!


> Rather I believe if you flip the document around you achieve something slightly more tolerable
> 
> ```
> (normal_mode)Escape niv
> (normal_mode)Ctrl+I n v
> ```
> Where the type annotation now denotes the mode

That's a command, not the mode, apologies for a confusing example, `normal_mode` here is `switch_to_normal_mode`. Also, there can be a sequence of commands, not just one, so they wouldn't fit in a single type (besides, they aren't a type, so that's also a category fit issue). Besides the command names can be rather long, so that would break the nice vertical alignment



> You could also use sub nodes instead to denote the modes to remove the visual clutter of the annotations
> 
> ```
> normal_mode {
>   Escape niv
>   Ctrl+I n v
> }
> ```

The plan was to have user-definable categories with the same mechanism of (types) to set modes per category, so (this particular example fails since mode switching would not have a common mode, but just as a bad illustration)

```
(n)"my mode switching commands" {
  // by default, only work in normal mode
  I "switch_to_insert_mode"
  V "switch_to_visual_mode"
  // but can also be overriden for any given keybind
  (i)Ctrl+I "switch_to_normal_mode"
}
```

**Comment by IceDragon200** [comment]
Hmm, I could see this getting out of hand really quick if we allowed multiple values in annotations, I'd like to get the feedback from the rest of the team on this one

**Comment by eugenesvk** [comment]
The idea was that it would still be parsed as a single value, though, so 
for `(   n v   )     Ctrl+X "cut_selection"` the type value that the parsed app gets is 
- `n v`, not
- `n` and `v`!

But it might indeed be too complicated to introduce another almost-quoted string-without-quotes-and-parentheses

**Comment by IceDragon200** [comment]
Yeah at that point you just quote the whole thing, a little annoying but it achieves what you'd need
```
// Since bare identifiers are allowed in kdl2 you can skip the quotes on the arguments now as long as it's an allowed identifier
("n v")Ctrl+X cut_selection
```

**Comment by zkat** [comment]
This is already integrated as part of kdl v2 and there was already a discussion about multiple values in them and we decided against it, so in closing this as a duplicate of that.

**Comment by tabatkins** [comment]
>  it would not allow whitespace within type values

Right, ident strings are terminated by whitespace, so you have to use one of the other string types if you want to write a type annotation with whitespace in it.

**Comment by eugenesvk** [comment]
> was already a discussion about multiple values in them

this is about a **single** value that allows using space without `""` since `()` can perform the same function of separating different elements instead of ` `

**Comment by IceDragon200** [comment]
The thing is the annotation spec is basically this:

```
annotation = "(" (ident | string) ")"
```

And idents don't allow spaces hence why it's not supported at all, you __need__ a string to have spaces in it

**Comment by eugenesvk** [comment]
Ok, so this could become
`annotation = "(" (ident | string | unquoted-string-with-unicode-spaces-because-you-don't-need-spaces-to-terminate) ")"`

**Comment by tabatkins** [comment]
In v2 the grammar is `'(' string ')'` (plus some whitespace, omitted for clarity) and idents are now just a type of string.

As was already stated, allowing *multiple* annotations on a value is intentionally not supported, and allowing a special-case syntax for writing an, I guess, "multi-part ident string" that only works in one spot isn't worthwhile. Better to have a simple, consistent grammar everywhere.
